File: apps/clojure/design-pattern/src/design_pattern/command.clj
(ns design-pattern.command)

;; Command パターン
;; コマンドをマップ {:execute fn, :undo fn, :description str} で表現する。

(defn make-command
  "コマンドを作成する。"
  [description execute-fn undo-fn]
  {:description description
   :execute     execute-fn
   :undo        undo-fn})

(defn execute
  "コマンドを実行する。"
  [cmd]
  ((:execute cmd)))

(defn undo
  "コマンドを取り消す。"
  [cmd]
  ((:undo cmd)))

;; --- ファイル操作コマンド ---

(defn create-file-command
  "ファイル作成コマンド（シミュレーション）。state-atom にファイルリストを保持。"
  [state-atom filename content]
  (make-command
    (str "Create file: " filename)
    (fn [] (swap! state-atom assoc filename content))
    (fn [] (swap! state-atom dissoc filename))))

(defn delete-file-command
  "ファイル削除コマンド。"
  [state-atom filename]
  (let [backup (atom nil)]
    (make-command
      (str "Delete file: " filename)
      (fn []
        (reset! backup (get @state-atom filename))
        (swap! state-atom dissoc filename))
      (fn []
        (when @backup
          (swap! state-atom assoc filename @backup))))))

;; --- コンポジットコマンド ---

(defn composite-command
  "複数のコマンドをまとめて実行するコンポジットコマンド。"
  [description commands]
  (make-command
    description
    (fn [] (doseq [cmd commands] (execute cmd)))
    (fn [] (doseq [cmd (reverse commands)] (undo cmd)))))

;; --- コマンド履歴 ---

(defn create-history
  "コマンド履歴を作成する。"
  []
  (atom []))

(defn execute-with-history!
  "コマンドを実行し、履歴に記録する。"
  [history cmd]
  (execute cmd)
  (swap! history conj cmd))

(defn undo-last!
  "最後に実行したコマンドを取り消す。"
  [history]
  (when-let [cmd (peek @history)]
    (undo cmd)
    (swap! history pop)))

File: apps/elixir/design-pattern/lib/design_pattern/command.ex
defmodule DesignPattern.Command do
  @moduledoc """
  Command パターン: 操作をマップとして表現し、実行・取り消しを可能にする。
  各コマンドは :execute と :undo のキーに関数を持つ。
  """

  @doc "コマンドを作成する"
  def new(name, execute_fn, undo_fn) do
    %{name: name, execute: execute_fn, undo: undo_fn}
  end

  @doc "コマンドを実行する"
  def execute(command, state) do
    command.execute.(state)
  end

  @doc "コマンドを取り消す"
  def undo(command, state) do
    command.undo.(state)
  end

  @doc "コマンド履歴を管理するヒストリーを作成する"
  def new_history do
    %{executed: [], state: nil}
  end

  @doc "コマンドを実行して履歴に追加する"
  def execute_with_history(history, command) do
    new_state = execute(command, history.state)
    %{history | executed: [command | history.executed], state: new_state}
  end

  @doc "最後のコマンドを取り消す"
  def undo_last(%{executed: []} = history), do: history

  def undo_last(%{executed: [last | rest]} = history) do
    new_state = undo(last, history.state)
    %{history | executed: rest, state: new_state}
  end

  @doc "マクロコマンド（複数コマンドの一括実行）"
  def macro(name, commands) do
    %{
      name: name,
      execute: fn state ->
        Enum.reduce(commands, state, fn cmd, acc -> cmd.execute.(acc) end)
      end,
      undo: fn state ->
        Enum.reduce(Enum.reverse(commands), state, fn cmd, acc -> cmd.undo.(acc) end)
      end
    }
  end
end
